tstest/clock.go

```diff
@@ -4,57 +4,495 @@
+	step time.Duration
+	// present is the last value returned by Now (and will be returned again by
+	// ConstantNow).
+	present time.Time
+	// skipStep indicates that the next call to Now should not add step to
+	// present. This occurs after initialization and after Advance.
+	skipStep bool
+	// timerChannelSize is the buffer size to use for channels created by
+	// NewTimer and NewTicker.
+	timerChannelSize int
+
+	events eventManager
+}
+
+func (c *Clock) init() {
+	if c.start.IsZero() {
+		c.start = time.Now()
+	}
+	if c.timerChannelSize == 0 {
+		c.timerChannelSize = 1
+	}
+	c.present = c.start
+	c.skipStep = true
+	c.events.AdvanceTo(c.present)
 }
 
 // Now returns the virtual clock's current time, and advances it
 // according to its step configuration.
 func (c *Clock) Now() time.Time {
-	c.Lock()
-	defer c.Unlock()
-	c.initLocked()
-	step := c.Step
-	ret := c.Present
-	c.Present = c.Present.Add(step)
-	return ret
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.skipStep && c.step > 0 {
+		c.present = c.present.Add(c.step)
+		c.events.AdvanceTo(c.present)
+	}
+	c.skipStep = false
+	return c.present
+}
+
+// ConstantNow returns the virtual clock's current time but does not advance it
+// if a step is configured.
+func (c *Clock) ConstantNow() time.Time {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.present
+}
+
+func (c *Clock) Advance(d time.Duration) time.Time {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.skipStep = true
+
+	if d > 0 {
+		c.present = c.present.Add(d)
+		c.events.AdvanceTo(c.present)
+	}
+
+	return c.present
+}
+
+func (c *Clock) Start() time.Time {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.start
+}
+
+func (c *Clock) Step() time.Duration {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.step
+}
+
+func (c *Clock) SetStep(d time.Duration) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.step = d
 }
 
-func (c *Clock) Advance(d time.Duration) {
-	c.Lock()
-	defer c.Unlock()
-	c.initLocked()
-	c.Present = c.Present.Add(d)
+func (c *Clock) SetTimerChannelSize(n int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.timerChannelSize = n
 }
 
-func (c *Clock) initLocked() {
-	if c.Start.IsZero() {
-		c.Start = time.Now()
+// NewTicker returns a Ticker that uses this Clock for accessing the current
+// time.
+func (c *Clock) NewTicker(d time.Duration) (tstime.TickerController, <-chan time.Time) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	t := &Ticker{
+		nextTrigger: c.present.Add(d),
+		period:      d,
+		em:          &c.events,
 	}
-	if c.Present.Before(c.Start) {
-		c.Present = c.Start
+	t.init(c.timerChannelSize)
+	return t, t.C
+}
+
+// NewTimer returns a Timer that uses this Clock for accessing the current
+// time.
+func (c *Clock) NewTimer(d time.Duration) (tstime.TimerController, <-chan time.Time) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	t := &Timer{
+		nextTrigger: c.present.Add(d),
+		em:          &c.events,
 	}
+	t.init(c.timerChannelSize)
+	return t, t.C
+}
+
+type eventHandler interface {
```

Comment:
docstrings for important unexported structs and interfaces. I loosely understand what eventHandler and event are doing here, but some additional docs (even if a bit boilerplatey) would be welcome to reduce uncertainty.

Reply:
Yeah, I really needed to do another pass to docstring all the things.

Now the only things without docstrings are pretty much a few unexported init receivers which should be obvious enough.